4.5 Exception Event

ADmad opened the pull request:
Allow returning a response from `Error.beforeRender` and `Exception.beforeRender` events. For `Exception.beforeRender` you can also switch the exception.

<!---

Please describe the reason for the changes you're proposing. If it is a large change, please open an issue to discuss first.

Always follow the [contribution guidelines](https://github.com/cakephp/cakephp/blob/master/.github/CONTRIBUTING.md) when submitting a pull request. In particular, make sure existing tests still pass, and add tests for all new behavior. When fixing a bug, you may want to add a test to verify the fix.

-->

ADmad committed:
Allow changing exception and returning a response using Exception.beforeRender event.

ADmad committed:
Allow returning response from Error.beforeRender event.

markstory commented on src/Error/ExceptionTrap.php:
```diff
@@ -363,7 +366,6 @@ public function logException(Throwable $exception, ?ServerRequestInterface $requ
                 $this->logger()->log($exception, $request);
             }
         }
-        $this->dispatchEvent('Exception.beforeRender', ['exception' => $exception]);
```
I think this was put here so that this event is also fired from the ErrorHandlerMiddleware. Because we needed to bridge both the ErrorHandler and ErrorRenderer interfaces the middleware operates ExceptionTrap at a lower level.

Right now I think this event will be missing in the ErrorHandlerMiddleware which is where most application exceptions are handled in.

ADmad replied:
If the `ErrorHandlerMiddleware` is configured to used the deprecated `ErrorHandler` then `ExceptionTrap` doesn't seem to be used at all, so I don't think the `Exception.beforeRender` event is fired in that case.

https://github.com/cakephp/cakephp/blob/630bd2d9749320e05d1ecadbdfa51d84d12a36e5/src/Error/Middleware/ErrorHandlerMiddleware.php#L154-L157

That said this change would cause the event to be not fired when an exception is caught by `ErrorHandlerMiddleware` even when `ExceptionTrap` is used. So I'll have to make changes to address that.

ADmad committed:
Refactor ErrorHandler::handleException().

This allows the hackish IntegrationTestTrait::disableErrorHandlerMiddleware()
to continue working.

ADmad commented:
The implementation of `IntegrationTestTrait::disableErrorHandlerMiddleware()` is some hacky shit.

markstory commented:
> The implementation of IntegrationTestTrait::disableErrorHandlerMiddleware() is some hacky shit.

Yes it is. I thought it was a reasonable amount of hackery given it is a debugging aid.
